src/util/wacz.ts

```diff
@@ -0,0 +1,424 @@
+  async generateToFile(filename: string) {
+    await pipeline(this.generate(), fs.createWriteStream(filename));
+  }
+
+  async *iterDirForZip(files: string[]): AsyncGenerator<InputWithoutMeta> {
+    const encoder = new TextEncoder();
+    const end = new EndMarker();
+    // correctly handles DST
+    const hoursOffset = (24 - new Date(0).getHours()) % 24;
+    const timezoneOffset = hoursOffset * 60 * 60 * 1000;
+    //const timezoneOffset = new Date().getTimezoneOffset() * 60000;
+
+    async function* wrapMarkers(
+      start: StartMarker,
+      iter: AsyncIterable<Uint8Array>,
+    ) {
+      yield start;
+      yield* iter;
+      yield end;
+    }
+
+    async function* getData(data: Uint8Array) {
+      yield data;
+    }
+
+    for (const filename of files) {
+      const input = fs.createReadStream(filename);
+
+      const stat = await fsp.stat(filename);
+      const mtime = stat.mtime;
+      const size = stat.size;
+
+      const nameStr = filename.slice(this.collDir.length + 1);
+      const name = encoder.encode(nameStr);
+      const lastModified = new Date(mtime.getTime() + timezoneOffset);
+
+      const start = new StartMarker(filename, nameStr, size);
+
+      yield { input: wrapMarkers(start, input), lastModified, name, size };
+    }
+
+    // datapackage.json
+
+    const datapackageData = encoder.encode(
+      JSON.stringify(this.datapackage, null, 2),
+    );
+
+    yield {
+      input: getData(datapackageData),
+      lastModified: new Date(),
+      name: DATAPACKAGE_JSON,
+      size: datapackageData.length,
+    };
+
+    const hash =
+      "sha256:" + createHash("sha256").update(datapackageData).digest("hex");
+
+    // datapackage-digest.json
+
+    const digest: WACZDigest = {
+      path: DATAPACKAGE_JSON,
+      hash,
+    };
+
+    // Get Signature
+    if (this.signingUrl) {
+      const body = JSON.stringify({
+        hash,
+        created: this.datapackage.created,
+      });
+
+      const headers: Record<string, string> = {
+        "Content-Type": "application/json",
+      };
+
+      if (this.signingToken) {
+        headers["Authorization"] = this.signingToken;
+      }
+
+      try {
+        const response = await fetch(this.signingUrl, {
+          method: "POST",
+          headers,
+          body,
+        });
+        digest.signedData = await response.json();
+      } catch (e) {
+        logger.warn(
+          "Failed to sign WACZ, continuing w/o signature",
+          { ...formatErr(e) },
+          "wacz",
+        );
+      }
+    }
+
+    const digestData = encoder.encode(JSON.stringify(digest, null, 2));
+
+    yield {
+      input: getData(digestData),
+      lastModified: new Date(),
+      name: DATAPACKAGE_DIGEST_JSON,
+      size: digestData.length,
+    };
+  }
+}
+
+// Merge CDX
+export function addDirFiles(fullDir: string): string[] {
+  const files = fs.readdirSync(fullDir);
+  return files.map((name) => path.join(fullDir, name));
+}
+
+export async function mergeCDXJ(
```

Comment:
One difference from py-wacz I can spot is that the resulting `index.cdx` no longer has a first line of `!meta 0 {"format": "cdxj-gzip-1.0", "filename": "index.cdx.gz"}`, should be simple enough to add

Reply:
Granted this is possibly unnecessary but seems like a good practice?

Reply:
ah yeah, forgot about this, should re-add for consistency.

Reply:
Added for index.idx in https://github.com/webrecorder/browsertrix-crawler/pull/673/commits/d2c172038bc2b80a28ac9653948a53d8bcc43f51